### PR TITLE
Hud Image Culling By Z Level: Theft edition

### LINKED ISSUE #63872: Taking off a medhud (item) doesn't remove the HUD (screen element)
Reporting client version: 514.1568

## Round ID:
[176088](https://scrubby.melonmesa.com/round/176088)

## Testmerges:
- [[MDB IGNORE] Adds progression traitors. Refactors uplink code in its entirety](https://www.github.com/tgstation/tgstation/pull/63588)
- [makes hud images only show to mobs on the same z level to save on maptick](https://www.github.com/tgstation/tgstation/pull/63558)

## Reproduction:

- Equip medical hud
- Change z levels
- Remove medical hud
- Still have the health bar above your head

image

### LINKED ISSUE #63901: Antag huds consistently don't work
## Round ID:

[176166](https://scrubby.melonmesa.com/round/176166)
Most recent Manuel rounds?

## Testmerges:

- [makes hud images only show to mobs on the same z level to save on maptick](https://www.github.com/tgstation/tgstation/pull/63558)

## Reproduction:

All antag huds consistently don't work.
Admin antag hud shows no antagonists - no traitors, revolutionaries, cultists, changelings, nothing
Team antag huds don't show any team-members for anyone involved, causing a large amount of confusion (especially on revolutionary rounds)
 
This has happened on both multi-z maps (icebox) and non-multi-z maps (deltastation)

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request
I'm re-pring kyler's initial work (#63558), since he's busy at the moment, and I think it would be useful.
I've had to make a minor change to how alternate appearance huds manage their key (314e5f70175a860b45e182c1c42852ced3850523)
Not 100% sure if this is the right way to go about it, the type is honestly kinda a deadend, but it seemed sane. Feedback'd be nice.

The rest of this is kyler's, cause he's based.

makes hud images only show to players on the same z level and linked z levels above to optimize load on sendmaps a bit more. also cleans up the code a bit. eventually i want to make huds support currently inactive images to improve savings more but ill do that later
fixes https://github.com/tgstation/tgstation/issues/63872
fixes https://github.com/tgstation/tgstation/issues/63901

## Why It's Good For The Game

so theres this step of the SendMaps profiler called "Check HUD/image visual contents". each tick, SendMaps iterates through every clients client.screen and client.images and checks each one for vis_contents and looking through a lot of sendmaps profiles it seems to take around 10% of all sendmaps varying a fair bit. for some reason images take about a fifth as long as screen movables to scan in this step buuuut a few days ago i realized that every ghost seems to have
image
images. and all of this it seems like is from atom_hud datums. it turns out that every single ghost gets: around 2 images of every mob in the world, 1 image of every door in the world, 1 image of every airlock in the world, 1 image of every bot path in the world, among presumably a few other painful examples. this is because they spawn with the advanced medical, security, and worst of all diagnostic huds. so every atom in the world those huds apply to are added to every ghost clients images list and increase maptick. each ghost also has around 60 screen movables from plane masters and such, but 60 * 5 is just 300, so screen objects are getting outweighed here by an order of magnitude. luckily most of the time ghosts are a minority of players, but on long rounds with lots of dead players i suspect the higher amounts of them can push maptick per person above 0.5

so by restricting it to images only in the players z level i (on local so the raw image count is lower than itd be on live) lowered it to around 300 on tramstation upper and lower levels and 500 on lavaland (before it was about 1600 on local)

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Huds aren't constructed in the way kyler initially thought, which makes parts of this harder. will get back to you

**Comment by realthomas69** [comment]
this appears to have fucked cult action buttons, it only shows two at once

**Comment by LemonInTheDark** [comment]
What? This should not be testmerged

**Comment by LemonInTheDark** [comment]
lmao. Will fix that tomorrow, kyler put something in place to deal with non global huds, so we should be good to retm this once I clean things up

**Comment by iprice** [comment]
Ghost/admin HUDs not showing player job on Terry presently (shows the generic ? ), probably related to this

**Comment by LemonInTheDark** [comment]
Likely yes. I will look into it, thanks a ton.

**Comment by LemonInTheDark** [comment]
> Ghost/admin HUDs not showing player job on Terry presently (shows the generic ? ), probably related to this

Handled, damn you kapu

**Comment by TheBonded** [comment]
both of the linked issue reports are already closed, but the comments seem to suggest they arent fixed?

**Comment by LemonInTheDark** [comment]
They were testmerge bugs, and I am dumb

File: code/datums/hud.dm [hunk 1]
```diff
@@ -17,116 +21,356 @@ GLOBAL_LIST_INIT(huds, list(
...
 		else
-			for(var/atom/A in hudatoms)
-				remove_from_single_hud(M, A)
+			break
 
-/datum/atom_hud/proc/remove_from_hud(atom/A)
-	if(!A)
+///returns a list of all hud users in the given z level and linked upper z levels (because hud users in higher z levels can see below)
+/datum/atom_hud/proc/get_hud_users_for_z_level(z_level)
+	if(z_level <= 0)
```
Comment (Kylerace) [inline code/datums/hud.dm:113]: should combine this with the stack_trace() below

Comment (LemonInTheDark) [reply code/datums/hud.dm:113]: yeet

**Pull request merged**